src/languages/r.js

```diff
@@ -131,48 +146,84 @@ export default function(hljs) {
           {begin: "'", end: "'", relevance: 0}
         ],
       },
+
+      // Matching numbers immediately following punctuation and operators is
+      // tricky since we need to look at the character ahead of a number to
+      // ensure the number is not part of an identifier, and we cannot use
+      // negative look-behind assertions. So instead we explicitly handle all
+      // possible combinations of (operator|punctuation), number.
+      // TODO: replace with negative look-behind when available
+      // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+\.[0-9a-fA-F]*[pP][+-]?\d+i?/ },
+      // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+([pP][+-]?\d+)?[Li]?/ },
+      // { begin: /(?<![a-zA-Z0-9._])(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?[Li]?/ }
       {
         relevance: 0,
-        className: {
-          2: "number"
-        },
         variants: [
-          // TODO: replace with negative look-behind when available
-          // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+\.[0-9a-fA-F]*[pP][+-]?\d+i?/ },
-          // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+([pP][+-]?\d+)?[Li]?/ },
-          // { begin: /(?<![a-zA-Z0-9._])(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?[Li]?/ }
           {
-            // Special case: only hexadecimal binary powers can contain fractions.
+            scope: {
+              1: 'operator',
+              2: 'number'
+            },
+            match: [
+              OPERATORS_RE,
+              NUMBER_TYPES_RE
+            ]
+          },
+          {
+            scope: {
+              1: 'operator',
+              2: 'number'
+            },
             match: [
-              /[^a-zA-Z0-9._]/, // not part of an identifier
-              /0[xX][0-9a-fA-F]+\.[0-9a-fA-F]*[pP][+-]?\d+i?/
+              /%[^%]*%/,
+              NUMBER_TYPES_RE
             ]
           },
           {
+            scope: {
+              1: 'punctuation',
+              2: 'number'
+            },
             match: [
-              /[^a-zA-Z0-9._]/, // not part of an identifier
-              /0[xX][0-9a-fA-F]+(?:[pP][+-]?\d+)?[Li]?/
+              PUNCTUATION_RE,
+              NUMBER_TYPES_RE
             ]
           },
           {
+            scope: { 2: 'number' },
+            relevance: 0,
             match: [
-              /[^a-zA-Z0-9._]/, // not part of an identifier
-              /(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?[Li]?/
+              /[^a-zA-Z0-9._]|^/, // not part of an identifier, or start of document
+              NUMBER_TYPES_RE
             ]
           }
         ]
       },
+
+      // Operators/punctuation when they're not directly followed by numbers
       {
-        // infix operator
-        begin: '%',
-        end: '%'
+        scope: 'operator',
+        match: /<-/
```

Comment:
Please add the full assignment regex back or is there some reason this had to change?  More explicit is better than less typically.

Reply:
The previous assignment regex was only used as a relevance boost, and only matches a small subset of all possible assignments, namely a subset of all legal identifiers, plus whitespace, plus assign, plus whitespace.

We can’t use this for highlighting since it misses many cases:

* Legal identifiers not included in the subset used here
* Lack of whitespace before or after the assignment
* Assignment to a complex expression rather than an identifier (e.g., at its simplest, `x[1] <- 2`)

I’m actually not sure why this regex was so overly specific even for relevance boosting: was it to avoid clashing with other languages? At any rate, assignment expressions in R can be (and routinely are, in actual code) quite complex.

Reply:
Sorry, commit bf1dc80fecfba5d8cca6c5f63494bf139b197d44 included an unnecessary, redundant declaration, I’ve removed that in cb7a392ce026805205a3adfed295885a839e8766. The result should be alright: `<-` is now handled in all situations, but the signal boost is only applied with the previous situation (except using the full `IDENT_RE`).